### PR TITLE
Support JUnit XML test report

### LINKED ISSUE #328: Report unit tests properly in the prow dashboards
Tests aren't populated properly in the prow dashboard, need to be fixed.

Sample test report: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-ibmcloud/327/pull-cluster-api-provider-ibmcloud-test/1425618612315492352

### PR DESCRIPTION
**Pull request by Amulyam24**
**What this PR does / why we need it**:
This PR provides the ability to generate JUnit XML test reports and populate the prow dashboard accordingly for the unit tests.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #328 

**Specify your PR reviewers**:

/cc @gyliu513 @jichenjc @xunpan @mkumatag 

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by Amulyam24** [comment]
cc @Rajalakshmi-Girish

**Comment by mkumatag** [comment]
> @Amulyam24: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-cluster-api-provider-ibmcloud-test	[ee1889e](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/commit/ee1889e9a65fd48f26d793680593220fc550ffe3)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-ibmcloud/340/pull-cluster-api-provider-ibmcloud-test/1438794493162688512)	`/test pull-cluster-api-provider-ibmcloud-test`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-ibmcloud&pr=340). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3AAmulyam24). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues?q=is:issue+is:open) when you hit one in your PR.

seems like this failure is happening here as well @Amulyam24 @Rajalakshmi-Girish

**Comment by Rajalakshmi-Girish** [comment]
> 
> 
> > @Amulyam24: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> > Test name	Commit	Details	Rerun command
> > pull-cluster-api-provider-ibmcloud-test	[ee1889e](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/commit/ee1889e9a65fd48f26d793680593220fc550ffe3)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-ibmcloud/340/pull-cluster-api-provider-ibmcloud-test/1438794493162688512)	`/test pull-cluster-api-provider-ibmcloud-test`
> > [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-ibmcloud&pr=340). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3AAmulyam24). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues?q=is:issue+is:open) when you hit one in your PR.
> 
> seems like this failure is happening here as well @Amulyam24 @Rajalakshmi-Girish

@Amulyam24 This failure seems to occur when make lint takes more than 1minute to execute.
You may try including --timeout option at https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/Makefile#L116

**Comment by Rajalakshmi-Girish** [comment]
The default value is 1m as mentioned here https://golangci-lint.run/usage/configuration/#command-line-options

**Comment by Amulyam24** [comment]
After the initial test failure with timeout, the retest fails with `/bin/sh: 1: go-junit-report: not found`
I'll make the changes to add a higher timeout as well to avoid the issue.

**Comment by Amulyam24** [comment]
> > > @Amulyam24: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> > > Test name	Commit	Details	Rerun command
> > > pull-cluster-api-provider-ibmcloud-test	[ee1889e](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/commit/ee1889e9a65fd48f26d793680593220fc550ffe3)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-ibmcloud/340/pull-cluster-api-provider-ibmcloud-test/1438794493162688512)	`/test pull-cluster-api-provider-ibmcloud-test`
> > > [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=cluster-api-provider-ibmcloud&pr=340). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3AAmulyam24). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues?q=is:issue+is:open) when you hit one in your PR.
> > 
> > 
> > seems like this failure is happening here as well @Amulyam24 @Rajalakshmi-Girish
> 
> @Amulyam24 This failure seems to occur when make lint takes more than 1minute to execute.
> You may try including --timeout option at https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/main/Makefile#L116

Thanks @Rajalakshmi-Girish , will update it.

**Comment by mkumatag** [comment]
> After the initial test failure with timeout, the retests fails with `/bin/sh: 1: go-junit-report: not found`
> I'll make the changes to add a higher timeout as well to avoid the issue.

this may be because you just vendored but not really installed it. just see if you can install it like [here](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/blob/a5766245b5a6f5380d719380983933522649e469/Makefile#L119)

**Comment by Amulyam24** [comment]
@mkumatag, made the required changes, PTAL.

File: Makefile [hunk 1]
```diff
@@ -24,12 +24,16 @@ GO_INSTALL = ./scripts/go_install.sh
...
+GO_JUNIT_REPORT_VER := v0.9.1
+GO_JUNIT_REPORT_BIN := go-junit-report
+GO_JUNIT_REPORT := $(TOOLS_BIN_DIR)/$(GO_JUNIT_REPORT_BIN)-$(GO_JUNIT_REPORT_VER)
 
 STAGING_REGISTRY ?= gcr.io/k8s-staging-capi-ibmcloud
 REGISTRY ?= $(STAGING_REGISTRY)
 RELEASE_TAG ?= $(shell git describe --abbrev=0 2>/dev/null)
 PULL_BASE_REF ?= $(RELEASE_TAG) # PULL_BASE_REF will be provided by Prow
 RELEASE_ALIAS_TAG ?= $(PULL_BASE_REF)
+REPORT_NAME := $(JUNIT_REPORT_NAME)
```
Comment (mkumatag) [inline Makefile:36]: no default set if not passed

File: Makefile [hunk 2]
```diff
@@ -49,8 +53,8 @@ endif
 all: manager
 
 # Run tests
-test: generate fmt vet manifests
-	go test ./... -coverprofile cover.out
+test: generate fmt vet manifests $(GO_JUNIT_REPORT)
+	go test ./... -v 2>&1 | $(GO_JUNIT_REPORT) > $(REPORT_NAME).xml
```
Comment (mkumatag) [inline Makefile:57]: I'm wondering if we can make optional, generate xml only when mentioned(especially via CI scripts)

**Review by mkumatag — CHANGES_REQUESTED** [review]
looking at the test results, it just shows only one entry, not sure why?!

**Comment by mkumatag** [comment]
@Amulyam24 see if this helps for the ginkgo - https://onsi.github.io/ginkgo/#generating-junit-xml-output

**Comment by Amulyam24** [comment]
> looking at the test results, it just shows only one entry, not sure why?!

Yes, go-junit-report is only considering the test suite as a whole and not the individual test cases unlike using the ginkgo package. I'll look more on it.

> I'm wondering if we can make optional, generate xml only when mentioned(especially via CI scripts)

@mkumatag, are you mentioning about making it optional via the `scripts/ci-test.sh` or through prow yaml. I'll have to explore the possibility of making it optional with the current flow, since with prow, the presence of artifacts will always lead to generation of the XML reports.

**Comment by mkumatag** [comment]
> @mkumatag, are you mentioning about making it optional via the `scripts/ci-test.sh` or through prow yaml. I'll have to explore the possibility of making it optional with the current PR flow, since with prow, the presence of artifacts will always lead to generation of the XML reports.

not with prow, e.g: if someone triggers these tests manually then it shouldn't generate the xml and generate the xml only if the option mentioned like some env like GENERATE_JUNIT, some of the logic can be derived from the kubernetes/kubernetes repo(something can be worked as followup PR).

**Comment by Amulyam24** [comment]
@mkumatag, updated with changes. PTAL.

**Review by mkumatag — APPROVED** [review]
/lgtm
/approve

**Pull request merged**